drivers/serial/uart_infineon_cat1.c

```diff
@@ -0,0 +1,467 @@
+
+/* Start processing interrupts in ISR.
+ * This function should be called the first thing in the ISR. Calling
+ * uart_irq_rx_ready(), uart_irq_tx_ready(), uart_irq_tx_complete()
+ * allowed only after this. */
+static int uart_cat1_irq_update(const struct device *dev)
+{
+	int status = 1;
+
+	if (((uart_cat1_irq_is_pending(dev) & CY_SCB_RX_INTR) != 0u) &&
+	    (Cy_SCB_UART_GetNumInRxFifo(DEV_DATA(dev)->obj.base) == 0u)) {
+		status = 0;
+	}
+
+	return status;
+}
+
+static void uart_cat1_irq_callback_set(const struct device *dev,
+				       uart_irq_callback_user_data_t cb, void *cb_data)
+{
+	struct uart_cat1_data *data = DEV_DATA(dev);
+	cyhal_uart_t *uart_obj = &data->obj;
+
+	/* Store user callback info */
+	data->irq_cb = cb;
+	data->irq_cb_data = cb_data;
+
+	/* Register a uart general callback handler  */
+	cyhal_uart_register_callback(uart_obj, _uart_event_callback_irq_mode, (void *) dev);
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+
+///////////////////////////////////////////////////////////////////////////////
+//                     UART API STRUCTURE                                    //
+///////////////////////////////////////////////////////////////////////////////
+static const struct uart_driver_api uart_cat1_driver_api = {
+	.poll_in = uart_cat1_poll_in,
+	.poll_out = uart_cat1_poll_out,
+	.err_check = uart_cat1_err_check,
+
+#ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
+	.configure = uart_cat1_configure,
+	.config_get = uart_cat1_config_get,
+#endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_cat1_fifo_fill,
+	.fifo_read = uart_cat1_fifo_read,
+	.irq_tx_enable = uart_cat1_irq_tx_enable,
+	.irq_tx_disable = uart_cat1_irq_tx_disable,
+	.irq_tx_ready = uart_cat1_irq_tx_ready,
+	.irq_rx_enable = uart_cat1_irq_rx_enable,
+	.irq_rx_disable = uart_cat1_irq_rx_disable,
+	.irq_tx_complete = uart_cat1_irq_tx_complete,
+	.irq_rx_ready = uart_cat1_irq_rx_ready,
+	.irq_err_enable = uart_cat1_irq_err_enable,
+	.irq_err_disable = uart_cat1_irq_err_disable,
+	.irq_is_pending = uart_cat1_irq_is_pending,
+	.irq_update = uart_cat1_irq_update,
+	.irq_callback_set = uart_cat1_irq_callback_set,
+#endif    /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+#ifdef CONFIG_UART_ASYNC_API
+	.callback_set = uart_cat1_async_cb_set,
+	.tx = uart_cat1_async_tx,
+	.tx_abort = uart_cat1_async_tx_abort,
+	.rx_enable = uart_cat1_async_rx_enable,
+	.rx_disable = uart_cat1_async_rx_disable,
+	.rx_buf_rsp = uart_cat1_async_rx_buf_rsp,
+#endif  /* CONFIG_UART_ASYNC_API */
+};
+
+///////////////////////////////////////////////////////////////////////////////
+//                     UART DRIVER INIT MACROS                               //
+///////////////////////////////////////////////////////////////////////////////
+#define INFINEON_CAT1_UART_INIT(n)						  \
+	static struct uart_cat1_data INFINEON_CAT1_uart##n##_data = { 0 };	  \
+										  \
+	static int INFINEON_CAT1_uart##n##_init(const struct device *dev)	  \
```

Comment:
why define a custom init function for each instance? please store DT config to driver's config and create a single one.